src/tls/test-server.c

```diff
@@ -194,51 +196,47 @@ assert_https_outcome (TestCase *tc,
       g_assert_cmpint (gnutls_credentials_set (session, GNUTLS_CRD_CERTIFICATE, xcred), ==, GNUTLS_E_SUCCESS);
 
       ret = gnutls_handshake (session);
-      if (expected_handshake_result == GNUTLS_E_SUCCESS)
+      if (ret != GNUTLS_E_SUCCESS)
         {
-          const gnutls_datum_t *server_certs;
-          unsigned server_certs_len;
-          ssize_t len;
-
-          if (ret != GNUTLS_E_SUCCESS)
+          if (expect_tls_failure)
+            exit (0);
+          else
             g_error ("Handshake failed: %s", gnutls_strerror (ret));
+        }
 
-          /* check server certificate */
-          server_certs = gnutls_certificate_get_peers (session, &server_certs_len);
-          g_assert (server_certs);
-          g_assert_cmpuint (server_certs_len, ==, expected_server_certs);
-
-          /* send request, read response */
-          g_assert_cmpint (gnutls_record_send (session, request, sizeof (request)), ==, sizeof (request));
-          len = gnutls_record_recv (session, buf, sizeof (buf) - 1);
-          if (expected_read_error != 0)
-            {
-              g_assert_cmpint (len, ==, expected_read_error);
-              close (fd);
-              exit (0);
-            }
-
-          g_assert_cmpint (len, >=, 100);
-          buf[len] = '\0'; /* so that we can use string functions on it */
-          /* This succeeds (200 OK) when building in-tree, but fails with dist-check due to missing doc root */
-          if (strstr (buf, "200 OK"))
-            {
-              cockpit_assert_strmatch (buf, "HTTP/1.1 200 OK\r\n"
-                                            "Content-Type: text/html\r\n"
-                                            "Content-Security-Policy: connect-src 'self' https://localhost wss://localhost;*");
-            }
-          else
-            {
-              cockpit_assert_strmatch (buf, "HTTP/1.1 404 Not Found\r\nContent-Type: text/html*");
-            }
+      /* check server certificate */
+      server_certs = gnutls_certificate_get_peers (session, &server_certs_len);
+      g_assert (server_certs);
+      g_assert_cmpuint (server_certs_len, ==, expected_server_certs);
 
-          g_assert_cmpint (gnutls_bye (session, GNUTLS_SHUT_RDWR), ==, GNUTLS_E_SUCCESS);
+      /* send request, read response */
+      len = gnutls_record_send (session, request, sizeof (request));
```

Comment:
Oh, also: why the send?  And is it actually possible for this to fail in response to a certificate error after we've successfully received data from the other side?  Seems strange....

Reply:
I don't understand the "why the send" part of the question -- we first need to send some HTTP request to cockpit-ws before we can actually read anything.

I haven't seen the first send+read succeed with a broken certificate. This would be really strange.. But I added an extra assertion at the end that will make the test case fail if that ever happens. That way we know that something is seriously broken. (It's impredictable enough as it is already..)

Reply:
Oh indeed, I got this backwards.  Please ignore.

Did you ever see the `send()` fail?  That's the actual new code here: replacing an old assertion with allowing failure in the case of expected fails.

Reply:
Yes, that was actually the trigger for this PR in the first place -- see the [comment above](https://github.com/cockpit-project/cockpit/pull/12743#issuecomment-529506529).